### PR TITLE
Implement simple lottery.

### LINKED ISSUE #69: Simple Lottery Construction - Prove/Verify
Prove and verify functionality for simple lottery construction.

### PR DESCRIPTION
**Pull request by tolikzinovyev**
## Content

Implement simple lottery's prove and verify functions. This PR doesn't have tests. Let's see what property tests will test and if it still makes sense to add other tests.

## Pre-submit checklist

- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
- PR
    - [x] No clippy warnings in the CI
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- Documentation
    - [x] Update README file (if relevant)
    - [x] Update documentation website (if relevant)

## Issue(s)

Closes #69

### CODE REVIEW

File: src/simple_lottery/algorithm.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+pub(super) fn verify(setup: &Setup, proof: &Proof) -> bool {
+    (proof.element_sequence.len() as u64 == setup.proof_size)
+        && proof
+            .element_sequence
+            .iter()
+            .all(|&element| hash(setup, element))
+}
+
+fn hash(setup: &Setup, element: Element) -> bool {
```
Comment (rrtoledo) [inline src/simple_lottery/algorithm.rs:32]: Perhaps rename this "play_lottery"?

Comment (rrtoledo) [inline src/simple_lottery/algorithm.rs:32]: Make this function public (or public(super) with an additional function in the wrapper) to check that an individual element can be sent

**Review by rrtoledo — CHANGES_REQUESTED** [review]
LGTM!

I would just add (a couple of) function(s) for the node to send a proof and the aggregator to check an individual element either exposing in algo or creating in the wrapper.
Doing so, we would support decentralized settings,
- the node can check locally if it has to send an element
- the aggregator could verify a single element as soon as they arrive, push them on a local and vector and cast it as a Proof when succesfully verifying the last one needed

We either can have 2 functions, separating the node from the aggregator, or one use by both.

File: src/simple_lottery/algorithm.rs (hunk 1 above)
Comment (tolikzinovyev) [reply src/simple_lottery/algorithm.rs:32]: I tried to be consistent with `src/centralized_telescope/algorithm.rs`. Let's name it `lottery_hash`?

**Review by tolikzinovyev — COMMENTED** [review]
I only planned to create an interface for the centralized setting. Let's work on the decentralized interface separately? I would be a different wrapper according to design doc. @rrtoledo

File: src/simple_lottery/algorithm.rs (hunk 1 above)
Comment (curiecrypt) [reply src/simple_lottery/algorithm.rs:32]: For consistency with the `centralized_telescope` and also expressing the functionality `lottery_hash` is good.

**Comment by rrtoledo** [comment]
> I only planned to create an interface for the centralized setting. Let's work on the decentralized interface separately? I would be a different wrapper according to design doc. @rrtoledo

Sure let's do it in a separate PR.
As for the wrapper, we should reuse the existing code so we can see whether we should have 2 wrappers for centralised and decentralised settings in wrapper.rs or not.

**Review by curiecrypt — CHANGES_REQUESTED** [review]
- Are you planning to add/enhance public function docs in another PR?
- I would write a simple test function even though the protests will cover most of the cases.
- Also, I saw a typo in file `simple_lottery/init.rs` line 27: `bound_completess`. FYI :) 
LGTM 👍

File: src/simple_lottery/algorithm.rs (hunk 1 above)
Comment (tolikzinovyev) [reply src/simple_lottery/algorithm.rs]: done

**Comment by tolikzinovyev** [comment]
@curiecrypt 
> Are you planning to add/enhance public function docs in another PR?

There is an open issue to do this for all files. #65 Not sure if I will have time this week though.

> I would write a simple test function even though the protests will cover most of the cases.

I'm afraid that if I write a test, it would become obsolete when property tests are implemented. However, if they do not test something, I would be happy to revisit this!

> Also, I saw a typo in file `simple_lottery/init.rs` line 27: `bound_completess`. FYI :)

Thanks, will make a PR with a fix.

File: src/simple_lottery/wrapper.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+impl Wrapper {
+    /// Initialize ALBA with `Params`.
+    pub fn create(params: &Params) -> Self {
+        let setup = make_setup(params);
+        Self::create_unsafe(&setup)
+    }
+
+    /// This function is unsafe to use and should be avoided.
+    /// Initialize ALBA with `Setup`.
```
Comment (rrtoledo) [inline src/simple_lottery/wrapper.rs:22]:
```suggestion
    /// Initialize ALBA with unchecked `Setup`.
```

File: src/simple_lottery/wrapper.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+impl Wrapper {
+    /// Initialize ALBA with `Params`.
+    pub fn create(params: &Params) -> Self {
+        let setup = make_setup(params);
+        Self::create_unsafe(&setup)
+    }
+
+    /// This function is unsafe to use and should be avoided.
```
Comment (rrtoledo) [inline src/simple_lottery/wrapper.rs:21]:
```suggestion
    /// Use this function with caution, e.g. in tests or with trusted `Setup`.
```

**Review by curiecrypt — APPROVED** [review]
LGTM 👍

File: src/simple_lottery/wrapper.rs (hunk 2 above)
Comment (tolikzinovyev) [reply src/simple_lottery/wrapper.rs:22]: Let's do this in #65. I copied this comment from `src/centralized_telescope/wrapper.rs`.

File: src/simple_lottery/wrapper.rs (hunk 3 above)
Comment (tolikzinovyev) [reply src/simple_lottery/wrapper.rs:21]: Let's do this in #65. I copied this comment from `src/centralized_telescope/wrapper.rs`.

**Comment by tolikzinovyev** [comment]
@rrtoledo waiting for your LGTM as promised :)

**Review by rrtoledo — APPROVED** [review]
LGTM with changes agreed to do in #65

**Pull request merged**